skyportal/handlers/api/gcn.py

```diff
@@ -0,0 +1,203 @@
+# Inspired by https://github.com/growth-astro/growth-too-marshal/blob/main/growth/too/gcn.py
+
+import datetime
+import os
+import gcn
+import lxml
+import xmlschema
+from urllib.parse import urlparse
+
+from baselayer.app.env import load_env
+
+from sqlalchemy.orm.exc import NoResultFound
+
+from baselayer.app.access import auth_or_token
+from ..base import BaseHandler
+from ...models import (
+    DBSession,
+    GcnEvent,
+    GcnNotice,
+    GcnTag,
+    Localization,
+)
+from ...utils.gcn import get_dateobs, get_tags, get_skymap, get_contour
+
+_, cfg = load_env()
+
+
+class GcnHandler(BaseHandler):
+    @auth_or_token
+    def put(self):
+        """
+        ---
+        description: Ingest GCN xml file
+        responses:
+          200:
+            content:
+              application/json:
+                schema: Success
+          400:
+            content:
+              application/json:
+                schema: Error
+        """
+        data = self.get_json()
+        payload = data['xml']
+
+        schema = f'{os.path.dirname(__file__)}/../../utils/schema/VOEvent-v2.0.xsd'
+        voevent_schema = xmlschema.XMLSchema(schema)
+        if voevent_schema.is_valid(payload):
+            root = lxml.etree.fromstring(payload.encode('ascii'))
+        else:
+            raise Exception("xml file is not valid VOEvent")
+
+        dateobs = get_dateobs(root)
+
+        try:
+            event = GcnEvent.query.filter_by(dateobs=dateobs).one()
+        except NoResultFound:
+            event = DBSession().merge(GcnEvent(dateobs=dateobs))
+            DBSession().commit()
```

Comment:
Is there a reason you call `DBSession().commit()` throughout the handler logic? We generally try to put off committing until the end of a handler so that we don't flush any changes to the database before the entire logic of the handler is successful, so that we don't have scenarios when for example, the `GcnEvent` is pushed to the database but then the app goes down before the corresponding tags get added so you have end up with a partial update of the database state.

Reply:
Ok I think I still need one after Localization is pushed otherwise the query for it fails (because the get_contour is done separately).

Reply:
Is this commit really needed?

Reply:
If it is, I'd suggest replacing it with `self.verify_and_commit()` so that we're properly enforcing permissions throughout the handler body